### PR TITLE
move starts_with from Utils to DB -- where it is used

### PR DESCRIPTION
**Pull request by hannesm**
simplify and move set_fixed_string into single user in Formats

shrink the Utils module, avoid opening it in Formats and GUI (where it is not
used much).

this is on the road to address #12

### CODE REVIEW

File: lib/gUI.ml [hunk 1]
```diff
@@ -156,10 +155,10 @@ let connect ~domid () =
   let version = Cstruct.create sizeof_gui_protocol_version in
   set_gui_protocol_version_version version qubes_gui_protocol_version_linux;
   QV.send qv [version] >>= function
-  | `Eof -> Lwt.fail (error "End-of-file sending protocol version")
+  | `Eof -> Lwt.fail (Utils.error "End-of-file sending protocol version")
```
Comment (talex5) [inline lib/gUI.ml:158]: I think these can just be e.g.

```suggestion
  | `Eof -> Lwt.fail_with "End-of-file sending protocol version"
```

**Review by talex5 — COMMENTED** [review]
Looks OK, but I think we could just remove the whole module. e.g.

- `raise (Utils.error` -> `Fmt.failwith`
- `Lwt.fail (Utils.error` -> `Lwt.fail_with`
- `starts_with` -> `Astring.String.is_prefix` (or wait for OCaml 4.12)
- `return`/`fail` -> `Lwt.return`/`Lwt.fail`

**Comment by hannesm** [comment]
ok, that was my goal in the end. I'll wait for #57 to be merged, and rebase and revise this PR. I'm not eager to introduce a dependency (on astrng) just for `prefix` -- so I suggest to leave it in here until we drop < 4.12 support.

thanks for your review.

**Comment by hannesm** [comment]
I force-pushed to this branch -- the Utils module is gone now.

File: lib/dB.ml [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
 open Lwt.Infix
 open Formats.QubesDB
 
 let (>>!=) x f =
   x >>= function
   | `Ok y -> f y
-  | `Eof -> fail (Failure "qubesdb-agent: end-of-file from QubesDB!")
-  | `Error (`Unknown msg) -> fail (error "qubesdb-agent: %s" msg)
+  | `Eof -> Lwt.fail_with "qubesdb-agent: end-of-file from QubesDB!"
+  | `Error (`Unknown msg) -> Lwt.fail (Fmt.failwith "qubesdb-agent: %s" msg)
```
Comment (talex5) [inline lib/dB.ml:11]: `Fmt.failwith` raises an exception, so either remove the unreachable `Lwt.fail` or use e.g. `Fmt.strf` to generate the error.
```suggestion
  | `Error (`Unknown msg) -> Fmt.failwith "qubesdb-agent: %s" msg
```

File: lib/dB.ml [hunk 3]
```diff
@@ -58,19 +67,19 @@ let full_db_sync t =
...
   let rec loop () =
     recv t.vchan >>= function
-    | QDB_RESP_MULTIREAD, "", _ -> return `Done
+    | QDB_RESP_MULTIREAD, "", _ -> Lwt.return `Done
     | QDB_RESP_MULTIREAD, path, data ->
         Log.debug (fun f -> f "%S = %S" path data);
         t.store <- t.store |> KeyMap.add path data;
         loop ()
-    | ty, _, _ -> fail (error "Unexpected QubesDB message: %s" (qdb_msg_to_string ty)) in
+    | ty, _, _ -> Lwt.fail (Fmt.failwith "Unexpected QubesDB message: %s" (qdb_msg_to_string ty)) in
```
Comment (talex5) [inline lib/dB.ml:75]: Same here.

File: lib/rExec.ml [hunk 4]
```diff
@@ -194,14 +193,14 @@ let send_hello t =
...
+  | `Ok () -> Lwt.return_unit
 
 let recv_hello t =
   recv t >>= function
-  | `Eof -> fail (error "End-of-file waiting for msg_hello")
-  | `Ok (`Hello, resp) -> return (get_peer_info_version resp)
-  | `Ok (ty, _) -> fail (error "Expected msg_hello, got %ld" (int_of_type ty))
+  | `Eof -> Lwt.fail_with "End-of-file waiting for msg_hello"
+  | `Ok (`Hello, resp) -> Lwt.return (get_peer_info_version resp)
+  | `Ok (ty, _) -> Lwt.fail (Fmt.failwith "Expected msg_hello, got %ld" (int_of_type ty))
```
Comment (talex5) [inline lib/rExec.ml:203]: Same here.

File: lib/rExec.ml [hunk 5]
```diff
@@ -215,15 +214,15 @@ let with_flow ~ty ~domid ~port fn =
       Lwt.catch
         (fun () ->
           recv_hello client >>= function
-          | version when version < 2l -> fail (error "Unsupported qrexec version %ld" version)
+          | version when version < 2l -> Lwt.fail (Fmt.failwith "Unsupported qrexec version %ld" version)
```
Comment (talex5) [inline lib/rExec.ml:217]: Same here.

File: lib/rExec.ml [hunk 6]
```diff
@@ -248,12 +247,12 @@ let port_of_int i =
...
   let cmd = Cstruct.to_string cmd in
   if cmd.[String.length cmd - 1] <> '\x00' then
-    fail (error "Command not null-terminated")
+    Lwt.fail_with "Command not null-terminated"
   else (
     let cmd = String.sub cmd 0 (String.length cmd - 1) in
     match cmd |> split ':' with
-    | None -> fail (error "Missing ':' in %S" cmd)
-    | Some (user, cmd) -> return (user, cmd)
+    | None -> Lwt.fail (Fmt.failwith "Missing ':' in %S" cmd)
```
Comment (talex5) [inline lib/rExec.ml:254]: Same here.

File: lib/rExec.ml [hunk 7]
```diff
@@ -377,9 +376,9 @@ let connect ~domid () =
   let t = { t; clients = Hashtbl.create 4; counter = 0; } in
   send_hello t.t >>= fun () ->
   recv_hello t.t >>= function
-  | version when version < 2l -> fail (error "Unsupported qrexec version %ld" version)
+  | version when version < 2l -> Lwt.fail (Fmt.failwith "Unsupported qrexec version %ld" version)
```
Comment (talex5) [inline lib/rExec.ml:379]: Same here.

**Review by talex5 — COMMENTED** [review]
`Lwt.fail (Fmt.failwith` seems a bit odd, as `failwith` never returns.

**Comment by hannesm** [comment]
thanks @talex5, fixed in a subsequent commit. I just dropped the `Lwt.fail` call (due to `Lwt.fail` documentation: Whenever possible, it is recommended to use [raise exn] instead, as [raise] captures a backtrace, while [Lwt.fail] does not. If you call [raise exn] in a callback that is expected by Lwt to return a promise, Lwt will automatically wrap [exn] in a rejected promise, but the backtrace will have been recorded by the OCaml runtime. Use [Lwt.fail] only when you specifically want to create a rejected promise, to pass to another function,  or store in a data structure.)

**Comment by talex5** [comment]
Thanks!

**Pull request merged**